src/partition/ReceivedPartition.cpp

```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
-    for (int rank = 1; rank < utils::MasterSlave::getSize(); rank++) {
-      int localNumberOfVertices = -1;
-      utils::MasterSlave::_communication->receive(localNumberOfVertices, rank);
-      PRECICE_DEBUG("Rank " << rank << " has " << localNumberOfVertices << " vertices.");
-      slaveOwnerVecs[rank].resize(localNumberOfVertices, 0);
-
-      if (localNumberOfVertices != 0) {
-        PRECICE_DEBUG("Receive tags from slave rank " << rank);
-        utils::MasterSlave::_communication->receive(slaveTags[rank], rank);
-        utils::MasterSlave::_communication->receive(slaveGlobalIDs[rank], rank);
-        PRECICE_DEBUG("Rank " << rank << " has tags " << slaveTags[rank]);
-        PRECICE_DEBUG("Rank " << rank << " has global IDs " << slaveGlobalIDs[rank]);
-        bool atInterface = false;
-        utils::MasterSlave::_communication->receive(atInterface, rank);
-        if (atInterface)
-          ranksAtInterface++;
+    for (auto & neighborRank: sharedVerticesSendMap)
+    {
+      int receiveSize = 0;
+      utils::MasterSlave::_communication->receive(receiveSize, neighborRank.first);
+      if (receiveSize != 0)
+      {
+        std::vector<double> receivedSharedVertices;
+        utils::MasterSlave::_communication->receive(receivedSharedVertices, neighborRank.first);
+        sharedVerticesReceiveMap.insert(std::make_pair(neighborRank.first, receivedSharedVertices));
       }
     }
 
-    // Decide upon owners,
-    PRECICE_DEBUG("Decide owners, first round by rough load balancing");
-    PRECICE_ASSERT(ranksAtInterface != 0);
-    int localGuess = _mesh->getGlobalNumberOfVertices() / ranksAtInterface; // Guess for a decent load balancing
-    // First round: every slave gets localGuess vertices
-    for (int rank = 0; rank < utils::MasterSlave::getSize(); rank++) {
-      int counter = 0;
-      for (size_t i = 0; i < slaveOwnerVecs[rank].size(); i++) {
-        // Vertex has no owner yet and rank could be owner
-        if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
-          slaveOwnerVecs[rank][i]                 = 1; // Now rank is owner
-          globalOwnerVec[slaveGlobalIDs[rank][i]] = 1; // Vertex now has owner
-          counter++;
-          if (counter == localGuess)
-            break;
+    // #6: Second round filter according to the number of owned vertices
+
+    int minGlobalID = _mesh->vertices()[0].getGlobalIndex();
+
+    for(int i=0; i < sharedVerticesList.size(); i++)
+    {
+      bool owned = true;
+
+      for (auto & sharingRank : sharedVerticesReceiveMap)
+      {
+        std::vector<double> vec = sharingRank.second;
+        if( std::find(vec.begin(), vec.end(), sharedVerticesList[i]) != vec.end() )
+        {
+          if((counter > neighborRanksVertexCount[sharingRank.first]) ||
+             (counter == neighborRanksVertexCount[sharingRank.first] && utils::MasterSlave::getRank() > sharingRank.first) )
+          {
+            owned = false;
+          }
         }
```

Comment:
Could you please add some documentation on how you do the load balancing?
Why the `utils::MasterSlave::getRank() > sharingRank.first`?

Reply:
Documentation is added. 
This is for the case where a vertex can be shared between two ranks and they have same vertex count: the lower rank will own the vertex.